flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/restore/AbstractRocksDBRestoreOperation.java

```diff
@@ -0,0 +1,135 @@
+	protected final KeyGroupRange keyGroupRange;
+	protected final int keyGroupPrefixBytes;
+	protected final int numberOfTransferringThreads;
+	protected final CloseableRegistry cancelStreamRegistry;
+	protected final ClassLoader userCodeClassLoader;
+	protected final ColumnFamilyOptions columnOptions;
+	protected final DBOptions dbOptions;
+	protected final Map<String, StateColumnFamilyHandle> kvStateInformation;
+	protected final RocksDBNativeMetricMonitor nativeMetricMonitor;
+	protected final File instanceBasePath;
+	protected final File instanceRocksDBPath;
+	protected final String dbPath;
+	protected List<ColumnFamilyHandle> columnFamilyHandles;
+	protected List<ColumnFamilyDescriptor> columnFamilyDescriptors;
+	protected final StateSerializerProvider<K> keySerializerProvider;
+	protected final SortedMap<Long, Set<StateHandleID>> restoredSstFiles;
+	protected final Collection<KeyedStateHandle> restoreStateHandles;
+
+	protected RocksDB db;
+	protected long lastCompletedCheckpointId = -1L;
+	protected UUID backendUID = UUID.randomUUID();
+	protected ColumnFamilyHandle defaultColumnFamilyHandle;
+
+	protected AbstractRocksDBRestoreOperation(
+		KeyGroupRange keyGroupRange,
+		int keyGroupPrefixBytes,
+		int numberOfTransferringThreads,
+		CloseableRegistry cancelStreamRegistry,
+		ClassLoader userCodeClassLoader,
+		Map<String, StateColumnFamilyHandle> kvStateInformation,
+		StateSerializerProvider<K> keySerializerProvider,
+		File instanceBasePath,
+		File instanceRocksDBPath,
+		DBOptions dbOptions,
+		ColumnFamilyOptions columnOptions,
+		RocksDBNativeMetricMonitor nativeMetricMonitor,
+		@Nonnull Collection<KeyedStateHandle> stateHandles) {
+		this.keyGroupRange = keyGroupRange;
+		this.keyGroupPrefixBytes = keyGroupPrefixBytes;
+		this.numberOfTransferringThreads = numberOfTransferringThreads;
+		this.cancelStreamRegistry = cancelStreamRegistry;
+		this.userCodeClassLoader = userCodeClassLoader;
+		this.kvStateInformation = kvStateInformation;
+		this.keySerializerProvider = keySerializerProvider;
+		this.instanceBasePath = instanceBasePath;
+		this.instanceRocksDBPath = instanceRocksDBPath;
+		this.dbPath = instanceRocksDBPath.getAbsolutePath();
+		this.dbOptions = dbOptions;
+		this.columnOptions = columnOptions;
+		this.nativeMetricMonitor = nativeMetricMonitor;
+		this.restoreStateHandles = stateHandles;
+		this.restoredSstFiles = new TreeMap<>();
+		this.columnFamilyHandles = new ArrayList<>(1);
+		this.columnFamilyDescriptors = Collections.emptyList();
+	}
+
+	public void openDB() throws IOException {
+		db = RocksDBOperationUtils.openDB(dbPath, columnFamilyDescriptors,
+			columnFamilyHandles, columnOptions, dbOptions);
+		// remove the default column family which is located at the first index
+		defaultColumnFamilyHandle = columnFamilyHandles.remove(0);
+		if (nativeMetricMonitor != null) {
+			nativeMetricMonitor.setRocksDB(db);
```

Comment:
This looks not like a good idea, because it is indicating a wrong order of object creation. Why do we need to have the metrics already created before this RocksDB instance? Couln't we just create the metrics towards the end, before creating the `RocksDBKeyedStateBackend` and after we already have the `RocksDB` instance? Then we can remove the setter-style and pass the `RocksDB` object into the contructor of the metric.

Reply:
I think I see why this was done, because you need the metric when the KV states are registered. But can't we just create the metric in this class if required, after the DB was opened? then the builder can get the metrics with a getter before passing it into the `RocksDBKeyedStateBackend`.